WIP: sfdisk implementation

prestist opened the pull request:
This is a WIP PR to add sfdisk support to ignition, Current state is iterating through failing tests and adding mapping to pretend to interface the same as before with sgdisk. 

Resolves #1926

prestist commented:
I think I might need a bit more for wipefs to work correctly. (looking into it but wanted to share)

Clean device using old way
```
[core@cosa-devsh ~]$ sudo sgdisk --zap-all /dev/vda
Found valid GPT with corrupt MBR; using GPT and will write new
protective MBR on save.
GPT data structures destroyed! You may now partition the disk using fdisk or
other utilities.
```

Write to device using command I will use again to verify it works
```
[core@cosa-devsh ~]$ sudo sgdisk --new=1:0:+65536 --change-name=1:create-partition --typecode=1:B921B045-1DF0-41C3-AF44-4C6F280D3FAE --partition-guid=1:05AE8178-224E-4744-862A-4F4B042662D0 /dev/vda
Creating new GPT entries in memory.
The operation has completed successfully.
```
Wipe device using wipefs to see if we can use a device that is wiped with it

```
[core@cosa-devsh ~]$ sudo wipefs -a -t mbr -t gpt -t pmbr /dev/vda
/dev/vda: 2 bytes were erased at offset 0x000001fe (PMBR): 55 aa
/dev/vda: calling ioctl to re-read partition table: Success
```

Verify disk is empty according to sfdisk

```
[core@cosa-devsh ~]$ sudo sfdisk --list /dev/vda
Disk /dev/vda: 2 GiB, 2147483648 bytes, 4194304 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
```

Try sgdisk command to create partition created above.

```
[core@cosa-devsh ~]$ sudo sgdisk --new=1:0:+65536 --change-name=1:create-partition --typecode=1:B921B045-1DF0-41C3-AF44-4C6F280D3FAE --partition-guid=1:05AE8178-224E-4744-862A-4F4B042662D0 /dev/vda
Found valid GPT with corrupt MBR; using GPT and will write new
protective MBR on save.
Could not create partition 1 from 67584 to 133119
Error encountered; not saving changes.
```

prestist commented:
Hmm, I seem to be stumbling a bit, I ended up focusing directly on the pretend section today. I found a few odds and ends that were missed in the first pass. 

I had been running against a creation test which only created one partition, however I stumbled into a test which makes two partitions, this is relevant because the nature of how the sfdisk is currently receiving my script input is painful. Currently I don't see a way to express multiple partitions when running commands in the `echo <my script> | sudo sfdisk /dev/vda`

using '\n' does not do what I would hope, so I think that leaves me two two imput ideas which I see

1. use cmd.stdin() after running `cmd.execute(sfdisk /dev/vda)`
2. create a file and read it in `sfdisk /dev/vda < myfile.txt`

I have been leaning towards option 1 as it seems the least invasive, but I dont think its working as I expect, and its difficult to debug blackbox tests.  

Is option 2 even acceptable?

prestist commented:
Ok, in the latest push a few things have happened.

1. Pulled in some code from sgdisk into sfdisk to allow for easier testing aka sgdiskcommit() and sgdiskbuildoptions()
2. Focused on pretend, it looks like sfdisk is now pretending correctly based on the scripts provided, which is good, as I was [stuck](https://github.com/coreos/ignition/pull/1929#issuecomment-2330197442). The answer was `"%s"` my script so the new line characters would work properly. 
3. Found out that the code added for wipe is not sufficient, as it results in a currupt partition table. Spoke to @jlebon  and I am going to look into how --zap-all works with sgdisk. 

Next steps:

1. Found that partitions is doing some interpreting based on the pretend and it has an understanding of sgdisk today, need to expand that for sfdisk. 
2. Proceed into commit
4. Fix wipe, as sgdisk --zap-all is functional and the code in sfdisk using wipefs results in a corrupt partition table.

jlebon commented on internal/sfdisk/sfdisk.go:
```diff
@@ -0,0 +1,276 @@
+	// If wipe we need to reset the partition table
+	if op.wipe {
+		// sgdisk --zap-all /dev/sda destroys the partition table, let's do the same with wipefs
+		// Found out that wipefs is not sufficient, as it results in a corrupt partition table, look into sgdisk's way of doing --zap-all
```
Looks like https://github.com/util-linux/util-linux/issues/3191 is an outstanding RFE to have an `sgdisk --zap-all` like option for sfdisk. We should verify the sgdisk code, but according to https://github.com/util-linux/util-linux/issues/3191#issuecomment-2328936289, it should be sufficient to zero out the first and last 34 sectors of the disk.

jlebon replied:
Hmm but actually, testing locally I can't get the corrupted partition table message to happen after using `wipefs -a`. It also seems like `wipefs -a` does not delete filesystem data hosted at the partitions as I previously thought.

prestist replied:
Thank you; Yeah -a seems like the right combo for our needs. Thank you thank you.

jlebon commented on internal/sfdisk/sfdisk.go:
```diff
@@ -0,0 +1,276 @@
+	if op.wipe {
+		// sgdisk --zap-all /dev/sda destroys the partition table, let's do the same with wipefs
+		// Found out that wipefs is not sufficient, as it results in a corrupt partition table, look into sgdisk's way of doing --zap-all
+		cmd := exec.Command("sudo", distro.WipefsCmd(), "-t pmbr", "-t mbr", "-t gpt", "-a", op.dev)
```
```suggestion
		cmd := exec.Command("sudo", distro.WipefsCmd(), "-t", "pmbr,mbr,gpt", "-a", op.dev)
```

Though as per https://github.com/coreos/ignition/pull/1929/files#r1757502849 I'm starting to think we don't actually need `-t` and it'll do the right thing for us. By comparison, if I have an XFS filesystem directly on `/dev/vda` (not a partition) and I do `sgdisk --zap-all /dev/vda`, it'll happily nuke that data.

So then this would just be

```suggestion
		cmd := exec.Command("sudo", distro.WipefsCmd(), "-a", op.dev)
```

prestist replied:
100%, makes sense. Thank you for looking into that.

prestist commented:
Latest push includes the [parsing](https://github.com/coreos/ignition/pull/1929/commits/aaff066fdbdcd4c1d4929a9ac71c15520117b0e3#diff-3ddf9f3db954a788966f29667df68227b2622a67ab083eceeec94d237f96fd6fR296) for sfdisk's pretend.  Additionally added a few [unit](https://github.com/coreos/ignition/pull/1929/commits/aaff066fdbdcd4c1d4929a9ac71c15520117b0e3#diff-f7fbb45f66c4c49d62a1e78951368fe262a71f31ee621456795f0a2693b5d935) tests to sanity check the parsing. Will likely need to expand with examples but so far thats passing. 

Additionally adjusted the wipe script as per @jlebon 's comment. 

Next steps (from last comment):

1. ✅  Found that partitions is doing some interpreting based on the pretend and it has an understanding of sgdisk today, need to expand that for sfdisk. 
2. ☑️  Proceed into commit
3. ✅  Fix wipe, as sgdisk --zap-all is functional and the code in sfdisk using wipefs results in a corrupt partition table.

prestist commented on internal/exec/stages/disks/partitions.go:
```diff
@@ -20,28 +20,32 @@ package disks
-)
+func getDeviceManager(logger *log.Logger, dev string) device_managers.DeviceManager {
+	// To be replaced with build tag support or something similar.
+	if false {
```
Note: will be dynamic based on arguments later just hard coded switch atm.

prestist commented:
Update: added interface for device_managers (name could be better not sure what would be good; feedback wanted lol), thus abstracting implementation from partitions.go, additionally lifted and placed parse responsibility on the implementation rather then the caller. 

Still need to work against commit.

jlebon commented on internal/device_managers/deviceManagers.go:
```diff
@@ -0,0 +1,28 @@
+package device_managers
```
How about `partitioners`?

prestist replied:
Yeah thats a better name; will do!

prestist replied:
I know renames can be confusing in prs, so i did some commit juggling and it has been renamed; I hope I did it in a way that made sense.

jlebon commented on internal/device_managers/sfdisk/sfdisk.go:
```diff
@@ -0,0 +1,249 @@
+	}
+
+	script := op.buildOptions()
+	cmd := exec.Command("sh", "-c", fmt.Sprintf("echo -e \"%s\" | sudo %s --no-act %s", script, distro.SfdiskCmd(), op.dev))
```
Also, shouldn't need `sudo` here.

prestist replied:
Hmm, I was running into permissions issues for some reason; (I agree) but it was failing on permissions. Maybe because of the "sh" ?

prestist replied:
Ok, yeah is fine sudos go bye

jlebon commented on internal/device_managers/sfdisk/sfdisk.go:
```diff
@@ -0,0 +1,249 @@
+	}
+
+	// Prepare the data, and a regex for matching on partitions
+	partitionRegex := regexp.MustCompile(`^/dev/\S+\s+\S*\s+(\d+)\s+(\d+)\s+\d+\s+\S+\s+\S+\s+\S+.*$`)
```
This deserves some explanation.

Also, I feel like we should ignore anything before the "New situation:" line.

prestist replied:
Added some details

prestist commented on internal/device_managers/sfdisk/sfdisk.go:
```diff
@@ -0,0 +1,249 @@
+// ParseOutput takes the output from sfdisk. Similarly to sgdisk
+// it then uses regex to parse the output into understood values like 'start' 'size' and attempts
+// to catch any failures and wrap them to return to the caller.
+func (op *Operation) ParseOutput(sfdiskOutput string, partitionNumbers []int) (map[int]device_managers.Output, error) {
```
Yeah; 100%,  Ok will do and tag it here once done.

prestist commented on internal/device_managers/sfdisk/sfdisk.go:
```diff
@@ -0,0 +1,249 @@
+
+	op.logger.Info("running sfdisk with script: %v", script)
+	exec.Command("sh", "-c", fmt.Sprintf("echo label: gpt | sudo %s %s", distro.SfdiskCmd(), op.dev))
+	cmd := exec.Command("sh", "-c", fmt.Sprintf("echo \"%s\" | sudo %s %s", script, distro.SfdiskCmd(), op.dev))
```
I gave it a swing

prestist commented:
Ugh; I accidentally rebased too far back and effected external commits sigh; I will work on fixing that.

prestist committed:
distro: add sfdisk command and partitioner backend selection

Add SfdiskCmd() and PartitionerBackend() which reads
ignition.partitioner from the kernel cmdline. Defaults to sgdisk.

prestist committed:
internal/partitioners: add DeviceManager interface with sgdisk and sfdisk backends

Introduce DeviceManager interface abstracting partition table
operations. Move sgdisk into internal/partitioners/sgdisk. Add sfdisk
backend using script mode with read-modify-write approach.

prestist committed:
internal/exec/stages/disks: use DeviceManager interface for partitioning

Replace direct sgdisk usage with getDeviceManager(). Include matching
partitions in commit for whole-table backends, add waitForUdev after
wipe, and gate partx behind NeedsPartx().

prestist committed:
tests: support sfdisk fallback in validator

Add sfdisk fallback to getPartitionSet and getPartitionInfo when sgdisk
is unavailable. Add manual comparison script.

prestist committed:
distro: temporarily default to sfdisk for CI testing

Flip the default partitioner from sgdisk to sfdisk so blackbox tests
exercise the sfdisk backend. This commit should be reverted or replaced
with a proper compile-time default mechanism.

prestist committed:
internal/partitioners/sfdisk: rewrite to whole-table with correct fixes

Revert to whole-table script approach (which had only 8 failures in
CI) and apply all discovered fixes:

- Slice-based merge for auto-numbered partitions (not map)
- Regex matches any absolute path (not just /dev/)
- Sort: fixed-size partitions first, fill-remaining last (prevents
  overlap in --no-act)
- last-lba header in script for correct fill-remaining bounds
- ParseOutput correction: +1 size when partition ends at lastLBA-1
- ensureGPTLabel: create empty GPT on disks without one so
  getLastLBA can read the correct value from the header
- --wipe-partitions always on both Pretend and Commit
- WritesCompleteTable returns true

prestist committed:
retrigger CI
